twilight-gateway/src/lib.rs

```diff
@@ -51,5 +49,142 @@ pub use twilight_gateway_queue as queue;
+    );
+
+    create_iterator(
+        (u32::from(bucket_id)..total).step_by(concurrency.into()),
+        total,
+        config,
+        per_shard_config,
+    )
+}
+
+/// Create a iterator of shards.
+///
+/// Passing a primary config is required. Further customization of this config
+/// may be performed in the callback.
+///
+/// # Examples
+///
+/// Start 10 out of 10 shards and count them:
+///
+/// ```no_run
+/// use std::{collections::HashMap, env, sync::Arc};
+/// use twilight_gateway::{Config, Intents};
+///
+/// # #[tokio::main]
+/// # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+/// let token = env::var("DISCORD_TOKEN")?;
+///
+/// let config = Config::new(token.clone(), Intents::GUILDS);
+/// let shards = twilight_gateway::create_iterator(0..10, 10, config, |_, builder| builder.build());
+///
+/// assert_eq!(shards.len(), 10);
+/// # Ok(()) }
+/// ```
+///
+/// # Panics
+///
+/// Panics if `range` contains values larger than `total`.
+///
+/// Panics if loading TLS certificates fails.
+#[track_caller]
+pub fn create_iterator<F: Fn(ShardId, ConfigBuilder<Q>) -> Config<Q>, Q: Clone>(
+    numbers: impl ExactSizeIterator<Item = u32>,
+    total: u32,
+    config: Config<Q>,
+    per_shard_config: F,
+) -> impl ExactSizeIterator<Item = Shard<Q>> {
+    numbers.map(move |index| {
+        let id = ShardId::new(index, total);
+        let config = per_shard_config(id, ConfigBuilder::from(config.clone()));
+
+        Shard::with_config(id, config)
+    })
+}
+
+/// Create a range of shards from Discord's recommendation.
+///
+/// Passing a primary config is required. Further customization of this config
+/// may be performed in the callback.
+///
+/// Internally calls [`create_iterator`] with the values from [`GetGatewayAuthed`].
+///
+/// # Errors
+///
+/// Returns a [`StartRecommendedErrorType::Deserializing`] error type if the
+/// response body failed to deserialize.
+///
+/// Returns a [`StartRecommendedErrorType::Request`] error type if the request
+/// failed to complete.
+///
+/// # Panics
+///
+/// Panics if loading TLS certificates fails.
+///
+/// [`GetGatewayAuthed`]: twilight_http::request::GetGatewayAuthed
+#[cfg(feature = "twilight-http")]
+pub async fn create_recommended<F: Fn(ShardId, ConfigBuilder<Q>) -> Config<Q>, Q: Clone>(
+    client: &Client,
+    config: Config<Q>,
+    per_shard_config: F,
+) -> Result<impl ExactSizeIterator<Item = Shard<Q>>, StartRecommendedError> {
+    let request = client.gateway().authed();
+    let response = request.await.map_err(|source| StartRecommendedError {
+        kind: StartRecommendedErrorType::Request,
+        source: Some(Box::new(source)),
+    })?;
+    let info = response
+        .model()
+        .await
+        .map_err(|source| StartRecommendedError {
+            kind: StartRecommendedErrorType::Deserializing,
+            source: Some(Box::new(source)),
+        })?;
+
+    Ok(create_iterator(
+        0..info.shards,
```

Comment:
Not sure why clippy doesn't complain about the leading `0` here, AFAIR that used to be a lint. Either way is fine though

Reply:
`..info.shards` doesn't implement `ExactSizeIterator` (for some reason)